migrate to gojmx

cristianciutea opened the pull request.

roobre commented on src/connection/jmx.go:
```diff
@@ -0,0 +1,102 @@
+		return
+	}
+
+	defer func() {
+		if err != nil {
+			p.sem.Release(1)
+		}
+	}()
```
I think I get what you want to achieve here (release after this func is called if there's an error) but I think doing it this way is not very intuitive for the reader.

Is there a different way in which we could do this? e.g. locking the semaphore after we check that `gojmx.NewClient` returns successfully?

roobre reviewed (changes requested):
Left a few nits here and there. However, the untested singleton dispatcher a blocker for me in order to merge this. I'd love if we could rewrite it into something less global and have a small test for it.

roobre reviewed (approved):
Unblocking this now, but I'd love if someone else on @newrelic/coreint could give it some 👀 as well

alvarocabanas commented on src/client/producer_consumer_collection.go:
```diff
@@ -64,23 +64,26 @@ func ConsumerWorker(consumerChan <-chan *args.JMXHost, wg *sync.WaitGroup, i *in
-				jmxwrapper.JMXClose() // Close needs to be called even on a failed open to clear out any set variables
-				jmxwrapper.JMXLock.Unlock()
 				continue
 			}
```
I would also move this and in the producerWorker bellow Jmx bellow to a private getConsumerJMXConnection or ConsumerJMXConnect function, just to have it more readable.

alvarocabanas reviewed (approved):
Nice work! I just put a couple of nits just to separate a bit the concerns, but completely optional.

cristianciutea commented:
> Left a few nits here and there. However, the untested singleton dispatcher a blocker for me in order to merge this. I'd love if we could rewrite it into something less global and have a small test for it.

I've removed the global singleton of connection provider and added tests for it.
Also did some refactor to address other comments.

roobre commented:
LGTM, thanks for taking care of this!
